text/0089-tenant-token.md

```diff
@@ -0,0 +1,125 @@
+| Term               | Definition |
+|--------------------|------------|
+| Master Key         | This is the master key that allows the creation of other API keys. The master key is defined by the user when launching MeiliSearch. |
+| API Key            | API keys are stored and managed from the endpoint `/keys` by the master key holder. These are the keys used by the technical teams to interact with MeiliSearch at the level of the frontend/backend code. |
+| Tenant Token     | These tokens are not stored and managed by a MeiliSearch instance. They are generated for each end-user by the backend code from a MeiliSearch API Key. They are used by the end-users to only search the documents belonging to them. |
+| Multi-Tenancy      | By multi-tenancy, we mean that an end-user only accesses data belonging to him within an index shared with other end-users. |
+
+### IV. Personas
+
+| Persona | Role |
+|---------|------|
+| Mark    | Mark is a developer for a SaaS company. He will implement the code to communicate with MeiliSearch to solve technical/product needs. |
+| UserX   | UserX represents any end-user searching from the frontend interfaces provided by Anna and Mark's company. |
+
+### V. `Tenant Token` Explanations
+
+#### Summary Key Points
+
+- `Tenant tokens` are generated from a MeiliSearch parent `API key` on the user's backend code. They are meant to resolve multi-tenancy by restricting access to data within an index according to the criteria chosen by the team managing the MeiliSearch instance.
+- `Tenant tokens` cannot be less restrictive than their respective parent `API key` and can only be used for searching with a predefined forced filter field.
+- `Tenant tokens` are not stored and thus not retrievable on MeiliSearch. This is why we highly advise setting an expiration time on them for security reasons.
+
+#### Solving Multi-Tenancy with `Tenant tokens`
+
+![](https://i.imgur.com/J4jVe1n.png)
+
+Let's say that `Mark` is a developer for a SaaS platform. He would like to ensure that end-user can only access their documents at search time. **His database contains many users and he hopes to have many more in the future.**
+
+When a user registers, the backend code generates a `Tenant token` specifically for that end-user so they can only access their documents.
+
+The `filter` parameter is set to restrict the search for documents having an `user_id` attribute. This `filter` parameter is contained in the `Tenant token` payload and cannot be modified during the search by the end-user making the request. `filter` can be made of any valid filters. e.g. `user_id = 10 and category = Romantic`
+
+This `Tenant token` is generated from a parent `API key` used to cipher the `Tenant token`. On the MeiliSearch side, the payload of the `tenant token` defines what data the user is allowed to retrieve, in order words, MeiliSearch grants authorization based on the permission defined in the Tenant token payload.
+
+---
+
+#### Generating a `Tenant token`
```

Comment:
So, the ideal workflow is not immediately clear to me. I'm assuming the code below is front-end due to it being javascript, but of course correct me if I'm wrong.

After a developer has generated a token from a set of restrictions and an API key, what do they do with? I'm assuming they must send it to MeiliSearch. Is there a new search parameter to do that? Or does the token effectively replace an API key?

Either way, on the back-end, inside an instance: is there a new setting so an index only accepts queries with a tenant token? If not, wouldn't it be fairly easy for someone to bypass restrictions?

Reply:
Here is the workflow that I hope will help you to understand better:

## User Objective

The user makes a medical SaaS and has an index that contains all medical reports of patients (end-user). He wants to make sure that the end-users will not see the medical reports of other patients.

## Tenant Token to the rescue

1. When the user goes to register or log in, he will have an identifier given by the user's system.
2. This system will generate a token signed by a MeiliSearch API Key (to know at search time that this token has been issued by a "valid" entity (Having knowledge of a MeiliSearch API Key) for this specific end-user. This token will hold within its payload, a description of rules (`indexesPolicy` claim) to be used at search time to ensure that this user can only see his data within a shared index.
3. This token will be given to the frontend code in a transparent way for the end-user.

## Benefits of the solution

- Completely transparent for MeiliSearch. MeiliSearch does not have to store business logic for each end-user and even less to store tokens for each end-user. There could be millions of them.
- The user generates the tenant tokens at will. He does not have to make any specific network exchange with MeiliSearch for that.